File: Programmers/서버 증설 횟수.cpp
#include <bits/stdc++.h>
using namespace std;

#ifdef LOCAL_TEST
constexpr bool isLOCAL = true;
#else
constexpr bool isLOCAL = false;
#endif

#define debug if constexpr (isLOCAL) cerr

int solution(vector<int> players, int m, int k) {
    int answer = 0;
    queue<int> q;

    for (int i=0; i<players.size(); i++) {
        int p = players[i];
        int ns = p / m;

        while (!q.empty()) {
            if (q.front() == i) q.pop();
            else break;
        }
        
        int s = q.size();
        if (ns > s) {
            for (int j=0; j<ns-s; j++) {
                q.push(i+k);
                answer++;
            }
        }
    }

    return answer;
}

File: Programmers/서버 증설 횟수.js
function solution(players, m, k) {
  let answer = 0;

  const queue = [];

  players.forEach((p, t) => {
    while (queue.length > 0 && queue[0] <= t) {
      queue.shift();
    }

    while (p >= (queue.length + 1) * m) {
      queue.push(t + k);
      answer++;
    }
  });

  return answer;
}

const testCases = [
  [[0, 2, 3, 3, 1, 2, 0, 0, 0, 0, 4, 2, 0, 6, 0, 4, 2, 13, 3, 5, 10, 0, 1, 5], 3, 5],
  [[0, 0, 0, 10, 0, 12, 0, 15, 0, 1, 0, 1, 0, 0, 0, 5, 0, 0, 11, 0, 8, 0, 0, 0], 5, 1],
  [[0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 0, 5, 0, 2, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1], 1, 1],
];

testCases.forEach((testCase) => {
  console.log(solution(...testCase));
});

